### PR TITLE
Implement round-half-even float.fromhex in common float_ops

### PR DESCRIPTION
**Pull request by youknowone**
## Summary

`float.fromhex` delegated to `literal::float::from_hex`, which is backed by `hexf_parse`. That reader is meant for exact literals: it rejects any inexact hexadecimal input and cannot tell overflow apart from a parse error, so `float.fromhex` could not round and never raised `OverflowError`. `test_from_hex` was consequently marked `@unittest.expectedFailure`.

This adds a dedicated hex-float reader in `common::float_ops` that reproduces the coefficient/exponent parsing, round-half-even rounding, and overflow detection that `float.fromhex` requires, and routes the classmethod through it.

## Changes

- `common::float_ops::from_hex(&str) -> Result<f64, HexFloatError>`: round-half-even over the full exponent range (subnormals down to `0x1p-1074`), the `inf`/`nan` spellings, surrounding ASCII whitespace, a saturating exponent parse, and `Invalid`/`TooLong`/`Overflow` error kinds.
- `float.fromhex` now maps `Overflow → OverflowError` and `Invalid`/`TooLong → ValueError`, with the messages Python uses.
- Removed `@unittest.expectedFailure` from `test_from_hex`.
- Added unit tests: exact bit patterns (including round-half-even ties), `inf`/`nan` sign bits, whitespace, and the error kinds.

## Verification

- `from_hex` output was compared bit-for-bit against CPython's `float.fromhex` over 125,000 inputs (curated + fuzzed, including near-halfway rounding and the `DBL_MAX` boundary): 0 mismatches.
- `cargo test -p rustpython-common` and `python -m test test_float` pass, with `test_from_hex` now green.

— *commented by Claude*

## Summary by CodeRabbit

* **New Features**
  * `float.fromhex()` now accepts more Python-style hex float inputs, including `inf`/`nan`, optional signs, whitespace, and full binary exponent parsing.
  * Error handling is now more specific for invalid formats, overly long strings, and overflow.
* **Bug Fixes**
  * Improved correctness for hexadecimal float rounding and edge cases (including trailing whitespace and special values).
* **Tests**
  * Added unit coverage to validate parsing results and error mapping.
* **Chores**
  * Updated spellchecker word list; consolidated Python ASCII-whitespace detection for parsing.

### CODE REVIEW

File: crates/common/src/float_ops.rs [hunk 1]
```diff
@@ -279,6 +279,413 @@ pub fn round_float_digits(x: f64, ndigits: i32) -> Option<f64> {
...
+}
+
+/// '0'-'9' -> 0..9, 'a'-'f'/'A'-'F' -> 10..15, else -1.
+#[inline]
+fn hex_from_char(c: u8) -> i32 {
+    match c {
+        b'0'..=b'9' => (c - b'0') as i32,
+        b'a'..=b'f' => (c - b'a') as i32 + 10,
+        b'A'..=b'F' => (c - b'A') as i32 + 10,
+        _ => -1,
```
Comment (ShaharNaveh) [inline crates/common/src/float_ops.rs]: Is there a reason for not returning an Option/Result?

File: crates/common/src/float_ops.rs [hunk 2]
```diff
@@ -279,6 +279,413 @@ pub fn round_float_digits(x: f64, ndigits: i32) -> Option<f64> {
...
+
+/// Read byte at `i`, returning a NUL terminator past the end so that
+/// digit/sign/space scans stop at the string boundary.
+#[inline]
+fn hex_byte_at(bytes: &[u8], i: usize) -> u8 {
+    if i < bytes.len() { bytes[i] } else { 0 }
+}
+
+#[inline]
+fn is_py_space(c: u8) -> bool {
```
Comment (ShaharNaveh) [inline crates/common/src/float_ops.rs]:
```suggestion
const fn is_py_space(c: u8) -> bool {
```

I thought we already had a function for detecting a py_space

File: crates/common/src/float_ops.rs [hunk 3]
```diff
@@ -279,6 +279,413 @@ pub fn round_float_digits(x: f64, ndigits: i32) -> Option<f64> {
...
+}
+
+const DBL_MANT_DIG: i64 = 53;
+const DBL_MIN_EXP: i64 = -1021;
+const DBL_MAX_EXP: i64 = 1024;
+
+/// Read byte at `i`, returning a NUL terminator past the end so that
+/// digit/sign/space scans stop at the string boundary.
+#[inline]
+fn hex_byte_at(bytes: &[u8], i: usize) -> u8 {
```
Comment (ShaharNaveh) [inline crates/common/src/float_ops.rs]: Same thing, should it return an Option instead?

File: crates/common/src/float_ops.rs [hunk 4]
```diff
@@ -279,6 +279,413 @@ pub fn round_float_digits(x: f64, ndigits: i32) -> Option<f64> {
...
+fn hex_from_char(c: u8) -> i32 {
+    match c {
+        b'0'..=b'9' => (c - b'0') as i32,
+        b'a'..=b'f' => (c - b'a') as i32 + 10,
+        b'A'..=b'F' => (c - b'A') as i32 + 10,
+        _ => -1,
+    }
+}
+
+/// `t` must be an ASCII-lowercase literal. Returns true iff every byte of `t`
```
Comment (ShaharNaveh) [inline crates/common/src/float_ops.rs]:
```suggestion
/// `t` must be an ASCII-lowercase literal. Returns true if every byte of `t`
```

File: crates/common/src/float_ops.rs [hunk 5]
```diff
@@ -279,6 +279,413 @@ pub fn round_float_digits(x: f64, ndigits: i32) -> Option<f64> {
...
+    while ti < t.len() && hex_byte_at(bytes, si).to_ascii_lowercase() == t[ti] {
+        si += 1;
+        ti += 1;
+    }
+    ti == t.len()
+}
+
+/// Returns `(retval, endptr)`. When the text at `p` is not inf/nan, `retval` is
+/// `-1.0` and `endptr == p`.
+fn parse_inf_or_nan(bytes: &[u8], p: usize) -> (f64, usize) {
```
Comment (ShaharNaveh) [inline crates/common/src/float_ops.rs]: same question

File: crates/common/src/float_ops.rs [hunk 6]
```diff
@@ -279,6 +279,413 @@ pub fn round_float_digits(x: f64, ndigits: i32) -> Option<f64> {
...
+    } else {
+        s = p;
+        retval = -1.0;
+    }
+    (retval, s)
+}
+
+/// Correctly-rounded scalbn. Every call site scales an already-representable
+/// value, so the result is exact.
+fn ldexp(x: f64, mut n: i32) -> f64 {
```
Comment (ShaharNaveh) [inline crates/common/src/float_ops.rs]: can this be const?

File: crates/common/src/float_ops.rs [hunk 7]
```diff
@@ -279,6 +279,413 @@ pub fn round_float_digits(x: f64, ndigits: i32) -> Option<f64> {
...
+    let insane_bound = core::cmp::min(
+        DBL_MIN_EXP - DBL_MANT_DIG - i64::MIN / 2,
+        i64::MAX / 2 + 1 - DBL_MAX_EXP,
+    ) / 4;
+    if ndigits_total > insane_bound {
+        return Err(HexFloatError::TooLong);
+    }
+
+    // [p <exponent>]
+    let exp = if hex_byte_at(bytes, idx) == b'p' || hex_byte_at(bytes, idx) == b'P' {
```
Comment (ShaharNaveh) [inline crates/common/src/float_ops.rs]:
```suggestion
    let exp = if matches!(hex_byte_at(bytes, idx), b'p' | b'P') {
```

File: crates/common/src/float_ops.rs [hunk 8]
```diff
@@ -279,6 +279,413 @@ pub fn round_float_digits(x: f64, ndigits: i32) -> Option<f64> {
...
+        idx += 1;
+        let exp_start = idx;
+        if hex_byte_at(bytes, idx) == b'-' || hex_byte_at(bytes, idx) == b'+' {
```
Comment (ShaharNaveh) [inline crates/common/src/float_ops.rs]: same like above

File: crates/common/src/float_ops.rs [hunk 9]
```diff
@@ -279,6 +279,413 @@ pub fn round_float_digits(x: f64, ndigits: i32) -> Option<f64> {
...
+            idx += 1;
+        }
+        if !(b'0' <= hex_byte_at(bytes, idx) && hex_byte_at(bytes, idx) <= b'9') {
```
Comment (ShaharNaveh) [inline crates/common/src/float_ops.rs]: same like above

File: crates/common/src/float_ops.rs [hunk 10]
```diff
@@ -279,6 +279,413 @@ pub fn round_float_digits(x: f64, ndigits: i32) -> Option<f64> {
...
+            return Err(HexFloatError::Invalid);
+        }
+        idx += 1;
+        while b'0' <= hex_byte_at(bytes, idx) && hex_byte_at(bytes, idx) <= b'9' {
```
Comment (ShaharNaveh) [inline crates/common/src/float_ops.rs]: same like above

**Review by ShaharNaveh — CHANGES_REQUESTED** [review]

**Comment by youknowone** [comment]
Thanks for the review — addressed in 51b3c1a9c:

- **`byte_at` (was `hex_byte_at`)** now returns `Option<u8>` (`bytes.get(i).copied()`); call sites use `== Some(..)` / `matches!(.., Some(b'p' | b'P'))`, which also folds in the `matches!` suggestions.
- **`hex_from_char`** now returns `Option<u8>` and is `const fn`; the scans go through a small `hex_digit_at` peek-and-decode helper.
- **`parse_inf_or_nan`** now returns `Option<(f64, usize)>` (no more `-1.0` sentinel).
- **whitespace helper** — you're right that it already existed: the copy is `is_py_ascii_whitespace` in `sre_engine`. I hoisted a shared `pub const fn is_py_ascii_whitespace` into `common::str` and use it here. I deliberately left `sre_engine`'s private copy in place: `sre_engine` is a standalone low-level crate (it only depends on `wtf8`), and making it depend on `common` would pull the whole runtime (`malachite-bigint`, `icu`, …) into a published regex crate. Happy to have `sre_engine` delegate to the shared one instead if you'd prefer that tradeoff.
- **`ldexp`** is now `const fn`; the `case_insensitive_match` doc wording is fixed.
- Dropped the redundant `test_` prefix on the unit tests (this was the `clippy::redundant_test_prefix` Lint failure).

— *commented by Claude*

**Comment by youknowone** [comment]
Update on the whitespace helper (after some more thought on where it belongs): instead of `common::str`, the shared `is_py_ascii_whitespace` now lives in **`rustpython-wtf8`** and is used by both `common::float_ops` and `sre_engine`. `wtf8` is the string crate that both already depend on, so this adds **zero new dependency edges**, and `sre_engine`'s previous private copy is removed in favor of it. (`rustpython-literal` was considered but reverted, since it would have added a new dependency to `sre_engine`.) See `bd187bf91`.

For the record, on two related questions that came up: the hex-float parser can't be borrowed from `ruff` (Python source has no hex-float literals, so ruff's number lexer only handles hex/oct/bin integers + decimal floats — the only `fromhex` in ruff is vendored typeshed stubs), and `ldexp` isn't reused from `pymath` because `pymath` only exposes `ldexp_bigint(f64, BigInt)`, which would force a BigInt allocation per call for what is an `i32` exponent here — hence the small self-contained `scalbn`.

— *commented by Claude*

**Review by ShaharNaveh — APPROVED** [review]
thanks for addressing the review:)

**Comment by youknowone** [comment]
@ShaharNaveh Thanks! I forgot to draft the first version
